src/Point/PointToInterlockingConnectionB4R1Impl.cs

```diff
@@ -0,0 +1,153 @@
+        var client = new RastaClient(channel);
+        _logger.LogTrace("Connecting...");
+        _timeout = new CancellationTokenSource();
+        _timeout.CancelAfter(10000);
+        var metadata = new Metadata { { "rasta-id", _localRastaId } };
+        _currentConnection = client.Stream(metadata, cancellationToken: _timeout.Token);
+    }
+
+    public async Task<bool> InitializeConnection(PointState state)
+    {
+        _logger.LogTrace("Connected. Waiting for request...");
+        if (await ReceiveMessage<PointPdiVersionCheckCommand>() == null)
+        {
+            _logger.LogError("Unexpected message.");
+            return false;
+        }
+
+        var versionCheckResponse = new PointPdiVersionCheckMessage(_localId, _remoteId, PointPdiVersionCheckMessageResultPdiVersionCheck.PDIVersionsFromReceiverAndSenderDoMatch, /* TODO */ 0, 0, new byte[] { });
+        await SendMessage(versionCheckResponse);
+
+        if (await ReceiveMessage<PointInitialisationRequestCommand>() == null)
+        {
+            _logger.LogError("Unexpected message.");
+            return false;
+        }
+
+        var startInitialization = new PointStartInitialisationMessage(_localId, _remoteId);
+        await SendMessage(startInitialization);
+
+        var pointState = new B4R1PointStateImpl(state);
+        var initialPosition = new PointPointPositionMessage(_localId, _remoteId, pointState.PointPosition, pointState.DegradedPointPosition);
+        await SendMessage(initialPosition);
+
+        var completeInitialization = new PointInitialisationCompletedMessage(_localId, _remoteId);
+        await SendMessage(completeInitialization);
+        return true;
+    }
+
+    public async Task SendPointPosition(PointState state)
+    {
+        var pointState = new B4R1PointStateImpl(state);
+        var response = new PointPointPositionMessage(_localId, _remoteId, pointState.PointPosition, pointState.DegradedPointPosition);
+        await SendMessage(response);
+    }
+
+    async public Task SendTimeoutMessage()
+    {
+        var response = new PointTimeoutMessage(_localId, _remoteId);
+        await SendMessage(response);
+    }
+
+    public async Task<PointPosition?> ReceivePointPosition()
+    {
+        var message = await ReceiveMessage<PointMovePointCommand>();
+
+        return (message != null)? message.CommandedPointPosition switch
+        {
+            PointMovePointCommandCommandedPointPosition.SubsystemElectronicInterlockingRequestsARightHandPointMoving => PointPosition.RIGHT,
+            PointMovePointCommandCommandedPointPosition.SubsystemElectronicInterlockingRequestsALeftHandPointMoving => PointPosition.LEFT,
+        } : null;
+    }
+
+    public void Dispose()
+    {
+        _currentConnection?.Dispose();
+    }
+
+    private async Task SendMessage(Message message)
+    {
+        if (_currentConnection == null) throw new NullReferenceException("Connection is null. Did you call Connect()?");
```

Comment:
I'd intuitively use a `InvalidOperationException` instead of `NullReferenceException`.